Frameworks/CoreGraphics/CGContext.mm

```diff
@@ -1898,86 +1906,197 @@ void CGContextSetPatternPhase(CGContextRef context, CGSize phase) {
+
 /**
  * Helper method to render text with the given DWRITE_GLYPH_RUN.
  *
  * @parameter glyphRun DWRITE_GLYPH_RUN object to render
  */
-HRESULT __CGContext::DrawGlyphRun(const DWRITE_GLYPH_RUN* glyphRun, bool transformByGlyph /* default true */) {
-    RETURN_HR_IF(E_INVALIDARG, !glyphRun);
-
-    // DWrite will crash if we try to give it glyphs that are below this threshold
-    // Though this value is approximate, it is small enough to not be noticeable while still safe
-    static constexpr float c_glyphThreshold = 0.5f;
+HRESULT __CGContext::DrawGlyphRuns(GlyphRunData* glyphRuns, size_t runsCount, bool transformByGlyph /* default true */) {
+    RETURN_HR_IF(E_INVALIDARG, !glyphRuns || runsCount == 0);
+    auto& state = CurrentGState();
+    if ((state.textDrawingMode & kCGTextFillStrokeClip) == 0) {
+        // Nothing to draw
+        return S_OK;
+    }
 
     // Undo assumed inversion about Y axis
     CGAffineTransform textTransform = CGAffineTransformScale(textMatrix, 1.0, -1.0);
     CGAffineTransform deviceTransform = CGContextGetUserSpaceToDeviceSpaceTransform(this);
+    CGAffineTransform combinedTransform = CGAffineTransformConcat(textTransform, deviceTransform);
 
     // Snap the vertical baseline to the nearest pixel to avoid blurring on devices with vertical antialiasing
     textTransform.ty = std::round(textTransform.ty);
 
-    CGAffineTransform finalTextTransform = CGAffineTransformConcat(textTransform, deviceTransform);
-    if ((fabs(finalTextTransform.a * glyphRun->fontEmSize) <= c_glyphThreshold &&
-         fabs(finalTextTransform.b * glyphRun->fontEmSize) <= c_glyphThreshold) ||
-        (fabs(finalTextTransform.d * glyphRun->fontEmSize) <= c_glyphThreshold &&
-         fabs(finalTextTransform.c * glyphRun->fontEmSize) <= c_glyphThreshold)) {
-        TraceWarning(TAG, L"Glyphs too small to be rendered");
+    bool fastPathTextDrawing =
+        ((textTransform.a == 1.0f && fabs(textTransform.d) == 1.0f && textTransform.b == 0.0f && textTransform.c == 0.0f) ||
+         !transformByGlyph) &&
+        state.textDrawingMode == kCGTextFill;
+
+    size_t maxGlyphCount = std::max_element(glyphRuns, glyphRuns + runsCount, [](const GlyphRunData& left, const GlyphRunData& right) {
+                               return left.run->glyphCount < right.run->glyphCount;
+                           })->run->glyphCount;
+
+    // Shared between transformed glyph runs
+    std::unique_ptr<CGFloat[]> zeroAdvances(new CGFloat[maxGlyphCount]());
+
+    // DWrite will crash if we try to give it glyphs that are below this threshold
+    // Though this value is approximate, it is small enough to not be noticeable while still safe
+    static constexpr float c_glyphThreshold = 0.5f;
+    std::vector<GlyphRunData> runs;
+
+    // For cases where we need to create glyphOffsets and runs
+    std::vector<std::vector<DWRITE_GLYPH_OFFSET>> createdOffsets;
```

Comment:
Possibly, but adds unneeded memory use on the fast path, & we're not certain of the size we'll need